Proposal for flexible composition of scanner components

jonschumacher opened the pull request:
Continuation of #18.
I am mainly posting this in order to let you know what I am planning to do in order to not duplicate work.

I now have a running version of the flexible composition including tests. Running in this context means that all dummy devices except the robot can be instantiated from a TOML configuration file. Each device is referenced with a unique ID and can also be retrieved by its type in order to e.g. get a DAQ device for acquisition.

Obviously, I broke a lot of stuff and would like to put the pieces together again. I did it this way since I am pretty convinced that the flexible composition stuff breaks a lot of tests if we do it the other way around and then the tests would not help us anyway. Therefore here is an outline of what I am planning to do:

- [ ]  Solve #20; I will probably go along the MDF definition and create structs according to the `/acquisition/` group and its subgroups. We have to discuss that in the issue.
- [ ] Solve #19: After we have an interface definition I can make the DAQ dummy work including a change in phase and amplitude. A simple setup of one send channel, one receive channel and a feedback channel should be sufficient. I already have a very simple Langevin simulation in MATLAB which should be easy to adapt.
- [ ] Fix the measurement interface for simulations and implement tests for the simulated DAQ.
- [x] Build a simulated robot that can be linked with a simulated gaussmeter. This should enable testing of field measurements in software.
- [ ] Build up a Red Pitaya test setup which mirrors the above setup in real hardware. I will add a voltage divider with a capacitor, which should allow for a phase change.

If all theses points have been done, we should be able to merge this request. After that we can tackle #22.

tknopp commented:
Comments:
* In general: awesome. I like both the general proposal how to proceed but also the implementation looks very good. I have not checked every line but in particular what you have done with the scanner looks buch better than what I had.
* Regarding `getDevice(scanner, "my_daq_id")` I like the flexibility but in practice it is not so clear to me how to now the ID. For instance the GUI just wants "the" daq and the GUI does not know about IDs
* The Manifest files should not be checked in. Best is to put that into gitignore.
* I like the structure where we have a dedicated folder and put everything in there. This folder should probably not live in MPIMeasurements itself (we do this mostly because we are lazy). We at IBI have a folder `.mpilab` where we put these things. We probably could just call this `.mpi`. Julia 1.6 has also introduced some package configuration infrastructure I have not looked into yet.
* Regarding Unitful I have some doubts how much to use it since in the past it made me quite some problems. This is also incompatible with what we do in MPIFiles / MPIReco. In my point of view Unitful is only helpful on the surface between a user (i.e. command line, config file) and the library itself. But then one should get rid of the units and internally just have a standard, i.e. a length of 1.0 is one meter. 
* TransferFunction has been removed. Not 100% sure where to put this put it is definitely important and in use. For instance in an MPS, the transfer functions are applied. The transfer function should probably best be located within the scanner folder in the future.
* I think you have seen that I made the Dummy work and integrated the changes. This allowed me to increase code coverage from 3% to 10%. 
* I will try to not make larger changes in MPIMeasurements in the next weeks so that our branches do not diverge to much. I might need to make smaller changes to get our scanner going but this should be rather minor things.

jonschumacher commented:
- _In general: awesome. I like both the general proposal how to proceed but also the implementation looks very good. I have not checked every line but in particular what you have done with the scanner looks buch better than what I had._
Thanks!
- _Regarding getDevice(scanner, "my_daq_id") I like the flexibility but in practice it is not so clear to me how to now the ID. For instance the GUI just wants "the" daq and the GUI does not know about IDs_
I added some convenience functions using getDevices(scanner, dataType) (please note the plural) in 15b5de8
- _The Manifest files should not be checked in. Best is to put that into gitignore._
Done in 0c52587.
- _I like the structure where we have a dedicated folder and put everything in there. This folder should probably not live in MPIMeasurements itself (we do this mostly because we are lazy). We at IBI have a folder .mpilab where we put these things. We probably could just call this .mpi. Julia 1.6 has also introduced some package configuration infrastructure I have not looked into yet._
With addConfigurationPath(path::String) (https://github.com/MagneticParticleImaging/MPIMeasurements.jl/blob/15b5de8c0999d4f6925551a81fce90c671374b85/src/MPIMeasurements.jl#L61) you can add arbitrary search paths for scanners, so I guess every way we want itshould work out. I am e.g. using it for the test configurations.
- _Regarding Unitful I have some doubts how much to use it since in the past it made me quite some problems. This is also incompatible with what we do in MPIFiles / MPIReco. In my point of view Unitful is only helpful on the surface between a user (i.e. command line, config file) and the library itself. But then one should get rid of the units and internally just have a standard, i.e. a length of 1.0 is one meter._
Concerning the use of units in software I always think of my course on software engineering where we had the example of the Mars Climate Orbiter. It might make some use cases more complicated but at some parts removes the ability to (easily) shoot your own leg. So I don't know how to proceed here. I would like to keep it in until we find a case where we can't reach our goal easily without removing Unitful.

- _TransferFunction has been removed. Not 100% sure where to put this put it is definitely important and in use. For instance in an MPS, the transfer functions are applied. The transfer function should probably best be located within the scanner folder in the future._
I understood the comment in https://github.com/MagneticParticleImaging/MPIMeasurements.jl/blob/4ba483cd351431da07ad2d9b136e9f200bf2e20c/src/MPIMeasurements.jl#L42 in such a way that it is not used anymore since it now lives in MPIFiles.
- _I think you have seen that I made the Dummy work and integrated the changes. This allowed me to increase code coverage from 3% to 10%._
Yes, I have merged those changes into my branch and will re-add the tests when I have resolved some other issues.
- _I will try to not make larger changes in MPIMeasurements in the next weeks so that our branches do not diverge to much. I might need to make smaller changes to get our scanner going but this should be rather minor things._
Yes, that would be great, even though I fully understand that you want your head scanner to work and have to make changes. I will just continue merging them on a regular basis.

tknopp commented:
* You are right, the transfer function stuff can definitely be removed. I forgot that we moved that stuff to MPIFiles.
* `getDAQ` and `getDAQs` are perfect for my use case. That is exactly what I wanted.
* Regarding Unitful I am absolutely fine to explore this more but I wanted to give a warning that this can have severe consequences since units propagate. So if you start here, you also need it in MPIFiles. If you use it in MPIFiles you need it in MPIReco. Then you need it in ImageUtils. And finally we need it in MPIUI. So this is a massive project. And even more important we need to have defined rules where to put units on. Just look at the voltage vector should its elements by voltages? You will not be able plotting it anymore when you do this without stripping the units. But maybe its good that you try to explore where this goes. Then we can learn from that experience how to proceed.

tknopp commented on src/MPIMeasurements.jl:
```diff
@@ -19,7 +19,12 @@ using InteractiveUtils
 using Graphics: @mustimplement
 using Scratch
 using Mmap
-using Gtk
```
@jonschumacher: Hi Jonas: I think we should drop Gtk/Winston altogether from MPIMeasurements. It was really just there as a workaround, which is disabled anyway. So just remove it entirely on your branch.

jonschumacher replied:
I actually added it again in order to have some user interaction when requesting the user to confirm that now sample is in the bore (https://github.com/MagneticParticleImaging/MPIMeasurements.jl/pull/25/commits/7ea2e569766d8239cf65be5fce5655a45a8c0028). I also came to the conclusion that this might not be the best choice when dealing with the CI issues. Ideally we should have some user interaction hooks that can handle either console input or GUI input from MPIUI. I guess I will just try to finish my rework stuff and then we should have a meeting on how to proceed.

tknopp replied:
Ah ok, then just proceed. I just wanted to indicate that it needs not to be kept because of me.
